File: core/src/Wall.ts
import { Tile } from './Tile.js';
import type { Dragon, NumberSuit, Wind } from './types.js';

export class Wall {
  private tiles: Tile[];

  constructor(tiles: Tile[] = []) {
    this.tiles = tiles;
  }

  static generateTiles(): Tile[] {
    const tiles: Tile[] = [];
    const numberSuits: NumberSuit[] = ['man', 'pin', 'sou'];
    for (const suit of numberSuits) {
      for (let value = 1 as const; value <= 9; value++) {
        for (let i = 0; i < 4; i++) {
          tiles.push(new Tile({ suit, value: value as 1|2|3|4|5|6|7|8|9 }));
        }
      }
    }

    const winds: Wind[] = ['east', 'south', 'west', 'north'];
    for (const value of winds) {
      for (let i = 0; i < 4; i++) {
        tiles.push(new Tile({ suit: 'wind', value }));
      }
    }

    const dragons: Dragon[] = ['white', 'green', 'red'];
    for (const value of dragons) {
      for (let i = 0; i < 4; i++) {
        tiles.push(new Tile({ suit: 'dragon', value }));
      }
    }
    return tiles;
  }

  static createShuffled(random: () => number = Math.random): Wall {
    const tiles = Wall.generateTiles();
    // Fisher-Yates shuffle
    for (let i = tiles.length - 1; i > 0; i--) {
      const j = Math.floor(random() * (i + 1));
      [tiles[i], tiles[j]] = [tiles[j], tiles[i]];
    }
    return new Wall(tiles);
  }

  shuffle(random: () => number = Math.random): void {
    for (let i = this.tiles.length - 1; i > 0; i--) {
      const j = Math.floor(random() * (i + 1));
      [this.tiles[i], this.tiles[j]] = [this.tiles[j], this.tiles[i]];
    }
  }

  draw(): Tile | undefined {
    return this.tiles.pop();
  }

  /** Peek at a tile without removing it. 0 refers to the next tile to be drawn. */
  peek(index = 0): Tile | undefined {
    return this.tiles[this.tiles.length - 1 - index];
  }

  get count(): number {
    return this.tiles.length;
  }
}

File: core/wall.py
"""Tile wall management."""
from __future__ import annotations

import random
from dataclasses import dataclass, field
from typing import List

from .models import Tile


def create_standard_wall() -> list[Tile]:
    """Return a shuffled list containing the full set of 136 tiles."""
    tiles: list[Tile] = []
    for suit in ("man", "pin", "sou"):
        for value in range(1, 10):
            tiles.extend([Tile(suit=suit, value=value) for _ in range(4)])
    for value in range(1, 5):
        tiles.extend([Tile(suit="wind", value=value) for _ in range(4)])
    for value in range(1, 4):
        tiles.extend([Tile(suit="dragon", value=value) for _ in range(4)])
    random.shuffle(tiles)
    return tiles


@dataclass
class Wall:
    """Represents the live wall plus dead wall, wanpai, and dora indicators."""

    tiles: List[Tile] = field(default_factory=list)
    dead_wall: List[Tile] = field(default_factory=list)
    dora_indicators: List[Tile] = field(default_factory=list)
    wanpai_size: int = 14

    def __post_init__(self) -> None:
        if not self.tiles:
            self.reset()

    def reset(self) -> None:
        """Fill the wall with a freshly shuffled standard tile set."""
        self.tiles = create_standard_wall()
        # Reserve the last 14 tiles as the dead wall
        self.dead_wall = [self.tiles.pop() for _ in range(14)]
        # Reveal the 5th tile from the end of the dead wall as the
        # dora indicator
        if len(self.dead_wall) >= 5:
            self.dora_indicators = [self.dead_wall[-5]]
        else:
            self.dora_indicators = []

    @property
    def remaining_yama_tiles(self) -> int:
        """Tiles left that can still be drawn this hand."""
        return max(len(self.tiles) - self.wanpai_size, 0)

    @property
    def remaining_tiles(self) -> int:
        """Return the number of tiles left in the wall."""
        return len(self.tiles)

    def draw_tile(self) -> Tile:
        """Draw and return the next tile from the wall."""
        return self.tiles.pop()
